engine/src/main/java/org/hibernate/search/query/dsl/sort/impl/SortFieldStates.java

```diff
@@ -0,0 +1,268 @@
+		}
+		else if ( currentType == SortField.Type.SCORE ) {
+			sortField = new SortField( null, SortField.Type.SCORE, isAsc() );
+		}
+		else if ( currentType == SortField.Type.DOC ) {
+			sortField = new SortField( null, SortField.Type.DOC, isDesc() );
+		}
+		else if ( coordinates != null ) {
+			sortField = new DistanceSortField( coordinates, currentName, isDesc() );
+			if ( hasMissingValue() ) {
+				throw new SearchException( "Missing values substitutes are not supported for distance sorting yet" );
+			}
+		}
+		else if ( currentLatitude != null ) {
+			sortField = new DistanceSortField( currentLatitude, currentLongitude, currentName, isDesc() );
+			if ( hasMissingValue() ) {
+				throw new SearchException( "Missing values substitutes are not supported for distance sorting yet" );
+			}
+		}
+		else if ( currentStringNativeSortFieldDescription != null ) {
+			sortField = new NativeSortField( currentName, currentStringNativeSortFieldDescription );
+		}
+		else {
+			sortField = new SortField( currentName, currentType, isDesc() );
+		}
+		processMissingValue( sortField );
+		sortFields.add( sortField );
+		reset();
+	}
+
+	public void guessCurrentSortFieldType() {
+		this.currentType = getCurrentSortFieldTypeGuess();
+	}
+
+	private SortField.Type getCurrentSortFieldTypeGuess() {
+		FieldDescriptor fieldDescriptor = queryContext.getFactory()
+				.getIndexedTypeDescriptor( queryContext.getEntityType() )
+				.getIndexedField( currentName );
+		switch ( fieldDescriptor.getType() ) {
+			case SPATIAL:
+				throw new SearchException( "Field '" + currentName + "' is a spatial field."
+						+ " For spatial fields, use .byDistance() and not .byField()." );
+			case NUMERIC:
+				NumericFieldSettingsDescriptor nfd = fieldDescriptor.as( NumericFieldSettingsDescriptor.class );
+				switch ( nfd.encodingType() ) {
+					case DOUBLE:
+						return SortField.Type.DOUBLE;
+					case FLOAT:
+						return SortField.Type.FLOAT;
+					case LONG:
+						return SortField.Type.LONG;
+					case INTEGER:
+						return SortField.Type.INT;
+					case UNKNOWN:
+						break; // Fail below
+				}
+				break;
+			case BASIC:
+				return SortField.Type.STRING;
+		}
+
+		throw new SearchException( "Cannot guess the field type for field '" + currentName
+				+ "'. Use byField(String, Sort.Type) to provide the sort type explicitly." );
+	}
+
+	private void processMissingValue(SortField sortField) {
+		if ( currentMissingValue != null && sortField.getType() != null ) {
+			if ( sortField.getType() == SortField.Type.STRING || sortField.getType() == SortField.Type.STRING_VAL ) {
+				if ( currentMissingValue == MISSING_VALUE_LAST ) {
+					sortField.setMissingValue( SortField.STRING_LAST );
+				}
+				else if ( currentMissingValue == MISSING_VALUE_FIRST ) {
+					sortField.setMissingValue( SortField.STRING_FIRST );
+				}
+				else {
+					throw new SearchException( "Unsupported 'use(Object)' for the field type: '" + currentType + "'."
```

Comment:
You documented the limitation, not the why. We need to have the why explained somewhere so that we can remove this limitation when we go away from the Lucene API.

I would go for something along the lines of:
Lucene sort API limits this feature to numeric fields. As our implementation is currently based on the Lucene API underneath, this is only available for numeric fields for all the indexing services, Elasticsearch included.